base/compiler/typeinfer.jl

```diff
@@ -66,24 +69,70 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
 
 _time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base not yet available.
 
-# We keep a stack of the Timings for each of the MethodInstances currently being timed.
+"""
+    Core.Compiler.Timings.clear_and_fetch_timings()
+
+Return, then clear, the previously recorded type inference timings.
+
+This fetches a vector of all of the type inference timings that have _finished_ as of this call. Note
+that there may be concurrent invocations of inference that are still running in another thread, but
+which haven't yet been added to this buffer. Those can be fetched in a future call.
+"""
+function clear_and_fetch_timings()
+    # Pass in the type, since the C code doesn't know about our Timing struct.
+    ccall(:jl_typeinf_profiling_clear_and_fetch, Any, (Any, Any,),
+          _finished_timings, Vector{Timing})::Vector{Timing}
+end
+
+function finish_timing_profile(timing::Timing)
+    ccall(:jl_typeinf_profiling_push_timing, Cvoid, (Any, Any,), _finished_timings, timing)
+end
+
+# DO NOT ACCESS DIRECTLY. This vector should only be accessed through the
+# functions above. It is a buffer that lives in the Julia module only to be *rooted*
+# for GC, but all accesses to the vector must go through C code, in order to be
+# thread safe.
+const _finished_timings = Timing[]
+
+# We store a profiling stack for *each Task* as a task-local-storage variable, _timings.
+# This is a stack of the Timings for each of the MethodInstances currently being timed.
 # Since type inference currently operates via a depth-first search (during abstract
 # evaluation), this vector operates like a call stack. The last node in _timings is the
 # node currently being inferred, and its parent is directly before it, etc.
 # Each Timing also contains its own vector for all of its children, so that the tree
 # call structure through type inference is recorded. (It's recorded as a tree, not a graph,
 # because we create a new node for duplicates.)
-const _timings = Timing[]
+# You will see this accessed below as `task_local_storage(:_timings)`
+
+# ------- Task Local Storage -------
+# Reimplementation of Task Local Storage, since these functions aren't available yet
+# at this stage of bootstrapping.
+current_task() = ccall(:jl_get_current_task, Ref{Task}, ())
+task_local_storage() = get_task_tls(current_task())
+function get_task_tls(t::Task)
+    if t.storage === nothing
+        t.storage = IdDict()
+    end
+    return (t.storage)::IdDict{Any,Any}
```

Comment:
@vchuravy / @pchintalapudi / @kpamnany: I finally got around to testing this, and indeed it doesn't work. :(

It turns out that `Base.IdDict` and `Core.Compiler.IdDict` are not the same thing. 😢 

Does anyone know why that is? Why doesn't Base just reexport the IdDict from Core.Compiler?
I.e. here: https://github.com/JuliaLang/julia/blob/80aeebe032a89455482134e8a9f1d46b1f5f2b40/base/Base.jl#L165
Why doesn't it just reexport the IdDict from Core.Compiler?

Reply:
BUT anyway, the error we encounter is that we're not allowed to set the Task's `.storage` field to a Core.Compiler.IdDict; it's expected to be a Base.IdDict.

So is there just no way to use task-local storage from Core.Compiler?

Reply:
Possibly a more robust solution here, which might also last longer if we parallelize inference, would be to store this on the `AbstractInterpreter` or `InferenceState` objects? Does anyone know if that's plausible? We essentially want some object that lives for the lifetime of the inference invocation, and is local to this invocation, where we can store the stack of timers.

Reply:
Or we'd have to thread it through every function call, but that seems like possibly a nonstarter.

Reply:
> So is there just no way to use task-local storage from Core.Compiler?

Nobody needed this before it seems :) 

> Does anyone know why that is? Why doesn't Base just reexport the IdDict from Core.Compiler?

Likely to isolate the compiler code from Base.

Reply:
You can do:

```
if isdefined(Core, :Main)
  Core.Main.Base.get_task_tls(current_task())
end
```

to get around the bootstrapping issue.

Reply:
It is because the compiler is forbidden from accessing TLS. Since we hijack the current task to run codegen, it could corrupt the running task if it does anything to interact with it.

Reply:
(instead, you could perhaps replace the TLS on entry with one from Core.Compiler, and restore it on return)